### PR TITLE
[HZAI-73] Vector API Docs

### PR DESCRIPTION
**Pull request by yuce**
* Added the docs for Vector API classes.
* Synced the defaults of `IndexConfig` with the server-side.

### CODE REVIEW

File: hazelcast/vector.py
Comment (EugeneAbramchuk) [inline hazelcast/vector.py:117]: this flag is missing from the args docs above, let's add it?

Comment (EugeneAbramchuk) [inline hazelcast/vector.py:117]: this flag is missing from the args docs above, let's add it?

File: hazelcast/proxy/vector_collection.py
Comment (EugeneAbramchuk) [inline hazelcast/proxy/vector_collection.py:41]: Looks like a surviving debug char :)

File: hazelcast/vector.py
Comment (EugeneAbramchuk) [inline hazelcast/vector.py:25]: Do we have a vector types description somewhere? I was unable to find it in the java docs either

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [inline hazelcast/proxy/vector_collection.py:211]: there is no server default, I removed `-1` default from Python client in https://github.com/hazelcast/hazelcast-python-client/pull/679

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py:25]: `DENSE` is just a regular vector. `SPARSE` vectors would be an optimization which are not supported yet.

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:211]: I think we should have a server default.

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py:118]: Added at: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/86856148d6d318eb453ee672625196cf2ad11e8c

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py]: Removed at: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/86856148d6d318eb453ee672625196cf2ad11e8c

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:211]: I've set it to 10 by default for now.

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [inline hazelcast/proxy/vector_collection.py:249]: what does it mean "optional operation" in this context? It is used also for in other proxies

Comment (yuce) [reply hazelcast/proxy/vector_collection.py:249]: I took it from the `Map` documentation. I think it means the operation is no-op if the mapping is not present.

Comment (yuce) [reply hazelcast/proxy/vector_collection.py:249]: `delete` deletes a mapping, it's not related to the proxy lifetime.
Does `VectorCollection.delete` behave differently than `Map.delete` if the mapping doesn't exist?

Comment (k-jamroz) [reply hazelcast/proxy/vector_collection.py:249]: > Does VectorCollection.delete behave differently than Map.delete if the mapping doesn't exist?

no, it is the same . I think I misunderstood before: you meant mapping=entry, I was thinking mapping=IMap.

I am not sure how to concisely say that the operation does nothing if there is nothing to do, but this is not what I understand when I read "optional operation". I think "idempotent" would be more precise.

and why `delete` is "optional" while `remove` is not? because the latter returns a value? (BTW, `remove` is not idempotent, because repeated it will return different result).

Comment (yuce) [reply hazelcast/proxy/vector_collection.py:249]: As far as I see from the server-side code, the difference between `remove` and `delete` may be related to map stores. `remove` calls `mapDataStore.load(key)` even if a mapping is not found.

Comment (yuce) [reply hazelcast/proxy/vector_collection.py:249]: Btw, I don't think `delete` is idempotent either.
Apart from that, even calls like `get` are not truly idempotent in Hz.

**Comment by yuce** [comment]
@k-jamroz @EugeneAbramchuk Could you guys approve the PR if the points you raised were addressed sufficiently?

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [reply hazelcast/proxy/vector_collection.py:254]: vector collection does not support map stores, so there is no difference between `remove` and `delete` that I would call "optional operation". So I'd remove `optional operation` from vector docs not to introduce confusion.

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [inline hazelcast/proxy/vector_collection.py:35]:
```suggestion
    The vector collection can work both async (non-blocking) or sync (blocking).
```

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [inline hazelcast/proxy/vector_collection.py:70]:
```suggestion
                >>> my_vc.set(key, value)
```

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [inline hazelcast/proxy/vector_collection.py:81]:
```suggestion
            Document associated with key or ``None`` if there is no mapping
            for key
```

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [inline hazelcast/proxy/vector_collection.py:60]:
```suggestion
        """Returns Document for the specified key, or ``None`` if this VectorCollection
```

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [inline hazelcast/proxy/vector_collection.py:215]:
```suggestion
            limit: Limit the maximum number of Documents returned.
```

File: hazelcast/vector.py
Comment (k-jamroz) [inline hazelcast/vector.py:41]:
```suggestion
        value: The value associated with this Document.
```

File: hazelcast/proxy/vector_collection.py
Comment (srknzl) [inline hazelcast/proxy/vector_collection.py:37]:
```suggestion
    However, async calls return a ``Future`` and do not block execution.
```

File: hazelcast/proxy/vector_collection.py
Comment (srknzl) [inline hazelcast/proxy/vector_collection.py:39]: It can also have a callback function to handle it full async, so I believe we should give a link to the Future related documentation from here and that's enough.

"However, async calls return ``Future`` and do not block execution." and a link should be enough

File: hazelcast/proxy/vector_collection.py
Comment (srknzl) [inline hazelcast/proxy/vector_collection.py:181]: What does this mean? Isn't this true for all objects for all APIs of the Hazelcast client? We don't specify this for other APIs though. Can it be confusing here as extra?

File: hazelcast/proxy/vector_collection.py
Comment (srknzl) [inline hazelcast/proxy/vector_collection.py:187]: if it's absent, how it will return the old Document?

File: hazelcast/proxy/vector_collection.py
Comment (srknzl) [inline hazelcast/proxy/vector_collection.py:213]:
```suggestion
            include_value: Whether to return value attached to the Document.
            include_vectors: Whether to return vectors attached to the Document.
```
I think better, not strong

File: hazelcast/proxy/vector_collection.py
Comment (srknzl) [inline hazelcast/proxy/vector_collection.py:215]: Why not document the default is 10 here?

File: hazelcast/vector.py
Comment (srknzl) [reply hazelcast/vector.py:25]: @yuce Do we have python documentation for Type enum(I guess it's a Enum) members? We should have imo

File: hazelcast/vector.py
Comment (srknzl) [inline hazelcast/vector.py:54]: I wonder how to access all vectors in the List if user requires that

File: hazelcast/vector.py
Comment (srknzl) [inline hazelcast/vector.py:131]: I think we can make this an inner class in VectorCollection. Not sure if it appears in the python API reference if we do that. It does not have to appear, so we can make this internal if possible

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:35]: That's directly from the `IMap` docs, so I'll keep that for consistency.

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:37]: That's straight from the `IMap` docs.

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:39]: That's straight from the `IMap` docs.

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:181]: Took it straight from `Map.put_if_absent`.

Comment (yuce) [reply hazelcast/proxy/vector_collection.py:181]: `put_if_absent` checks for equality of the serialized binary object on the server side. Not all methods do that.

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:213]: I think both versions are fine.

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py:54]: It's `document.vectors`.

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py:99]: I think these names are pretty self-explanatory. We can have more in-depth docs, like when would you use Cosine vs Euclidean, but it's probably should be done in the main docs.

Comment (k-jamroz) [reply hazelcast/vector.py:99]: these are well known metrics

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py:131]: Why make it an inner class of VectorCollection?

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py:28]: Added some docs at: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/75ff67f5ca7811e094e2a3603b04e59f51eeceab

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:255]: I've added `optional operation` to the `vc.remove` method for consistency.

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py]: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/75ff67f5ca7811e094e2a3603b04e59f51eeceab

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py]: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/75ff67f5ca7811e094e2a3603b04e59f51eeceab

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py]: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/75ff67f5ca7811e094e2a3603b04e59f51eeceab

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:187]: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/75ff67f5ca7811e094e2a3603b04e59f51eeceab

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py]: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/75ff67f5ca7811e094e2a3603b04e59f51eeceab

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py]: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/75ff67f5ca7811e094e2a3603b04e59f51eeceab

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py]: Added the default `10` at:
https://github.com/hazelcast/hazelcast-python-
client/pull/678/commits/75ff67f5ca7811e094e2a3603b04e59f51eeceab

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py]: https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/afce5c2467ca49176d0c1d9b199e65669d2d7268

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [reply hazelcast/proxy/vector_collection.py]: shouldn't we DRY? default `limit` is clearly visible as default value of the method parameter

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [reply hazelcast/proxy/vector_collection.py:256]: for consistency with what? Vector colleciton is not IMap and "optional operation" does not mean anything in context of vector collection. We should remove it because it is only misleading and confusing in case of vector collection.

File: hazelcast/proxy/vector_collection.py
Comment (k-jamroz) [reply hazelcast/proxy/vector_collection.py:35]: IMO we do not need typo-level consistency, but we can keep it as is

**Review by k-jamroz — CHANGES_REQUESTED** [review]
Remove meaningless "optional operation" from vector collection docs

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py:256]: `VectorCollection` is closest to the `IMap. As we've discussed, "optional operation" means the operation becomes no-op if the mapping does not exist. The user wouldn't necessarily know whether VectorCollection uses a map store or not.

Feel free to send another PR to explain this better after this PR is merged.

Comment (k-jamroz) [reply hazelcast/proxy/vector_collection.py:256]: > "optional operation" means the operation becomes no-op if the mapping does not exist

this is completely not obvious meaning for me, do you think that our users will understand it that way? @shultseva @vbekiaris how do understand that?

also:

1. there are more operations that do not change state of the system if the collection does not exist, eg. `get`, `search`, `size`, `clear`
2. for all those cases it does not matter if the collection exists or not (strictly speaking if `DistributedObject` exists) but if the collection (or partition for `get`) is empty or not (collection that does not exist is empty by definition). And the result is logically obvious: if the collection is empty then removing from it does not change it
3. we want to cut ties between vector collection and IMap, there is no value is artificially making them similar

so this is a useless, confusing information.

File: hazelcast/vector.py
Comment (EugeneAbramchuk) [inline hazelcast/vector.py:99]: This definition is somehow very convoluted. I am sure it can be improved but I am unsure how.

I propose to start with 
`Metric is the measure of similarity which is used for vector indexing to enable searching.`

The docs team can likely come up with something better

File: hazelcast/vector.py
Comment (EugeneAbramchuk) [inline hazelcast/vector.py:118]: I would rather add some explanation what deduplication means, but there is no mention of it anywhere in the docs elsewhere... (neither in the java docs)

Comment (k-jamroz) [reply hazelcast/vector.py:118]: it is described in https://deploy-preview-1124--hardcore-allen-f5257d.netlify.app/hazelcast/5.5-snapshot/data-structures/vector-collections#configuration (not yet merged)

File: hazelcast/vector.py
Comment (yuce) [reply hazelcast/vector.py:99]: Not sure your suggestion is more understandable than the previous one :)

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py]: I changed it as follows:
https://github.com/hazelcast/hazelcast-python-client/pull/678/commits/6fb9acf06987c07ea7d93941221a316818aa8514

File: hazelcast/proxy/vector_collection.py
Comment (yuce) [reply hazelcast/proxy/vector_collection.py]: The defaults are explicitly mentioned in the documentation in some other data structure methods, like `Map.add_index`, so I think it's OK to keep it.

File: hazelcast/vector.py
Comment (EugeneAbramchuk) [reply hazelcast/vector.py:99]: I agree my suggestion is not an improvement, I was waiting for the docs team to comment. We can pick this up later and merge as is for now

**Review by EugeneAbramchuk — APPROVED** [review]

**Review by srknzl — APPROVED** [review]

**Review by k-jamroz — COMMENTED** [review]
my [comment](https://github.com/hazelcast/hazelcast-python-client/pull/678/files#r1647615569) about "optional operation" was not addressed, but if others are happy with it, I will not block merging this PR

**Review by k-jamroz — APPROVED** [review]

**Comment by srknzl** [comment]
> my [comment](https://github.com/hazelcast/hazelcast-python-client/pull/678/files#r1647615569) about "optional operation" was not addressed, but if others are happy with it, I will not block merging this PR

I also think we should remove it if it's not obvious and creates confusion. However, maybe we should do it in general regarding all python docstrings in the repo

**Comment by yuce** [comment]
It seems tests jars for snapshots are not produced anymore. Because of that the test runner fails.
I've raised: https://hazelcast.atlassian.net/browse/HZAI-101 for that.

**Pull request merged**
